spring-cloud-gcp-samples/spring-cloud-gcp-bigquery-sample/src/main/java/com/example/WebController.java

```diff
@@ -49,6 +54,99 @@ public ModelAndView renderIndex(ModelMap map) {
+  @PostMapping("/uploadJsonFile")
+  public ModelAndView handleJsonFileUpload(
+      @RequestParam("file") MultipartFile file,
+      @RequestParam("tableName") String tableName,
+      @RequestParam(name = "createTable", required = false) String createDefaultTable)
+      throws IOException {
+    ListenableFuture<WriteApiResponse> writeApiRes = null;
+    if (createDefaultTable != null
+        && createDefaultTable.equals("createTable")) { // create the default table
+      writeApiRes =
+          this.bigQueryTemplate.writeJsonStream(
+              tableName, file.getInputStream(), getDefaultSchema());
+    } else { // we are expecting the table to be already existing
+      writeApiRes = this.bigQueryTemplate.writeJsonStream(tableName, file.getInputStream());
+    }
+    return getWriteApiResponse(writeApiRes, tableName);
+  }
+
+  private Schema getDefaultSchema() {
+    return Schema.of(
+        Field.of("CompanyName", StandardSQLTypeName.STRING),
+        Field.of("Description", StandardSQLTypeName.STRING),
+        Field.of("SerialNumber", StandardSQLTypeName.NUMERIC),
+        Field.of("Leave", StandardSQLTypeName.NUMERIC),
+        Field.of("EmpName", StandardSQLTypeName.STRING));
+  }
+
+  /**
+   * Handles JSON data upload using using {@link BigQueryTemplate}.
+   *
+   * @param jsonRows the String JSON data to upload to BigQuery
+   * @param tableName name of the table to load data into
+   * @return ModelAndView of the response the send back to users
+   */
+  @PostMapping("/uploadJsonText")
+  public ModelAndView handleJsonTextUpload(
+      @RequestParam("jsonRows") String jsonRows,
+      @RequestParam("tableName") String tableName,
+      @RequestParam(name = "createTable", required = false) String createDefaultTable) {
+    ListenableFuture<WriteApiResponse> writeApiRes = null;
+    if (createDefaultTable != null
+        && createDefaultTable.equals("createTable")) { // create the default table
+
+      System.out.println("Create table and write JSON Text " + createDefaultTable);
+      writeApiRes =
+          this.bigQueryTemplate.writeJsonStream(
+              tableName, new ByteArrayInputStream(jsonRows.getBytes()), getDefaultSchema());
+    } else { // we are expecting the table to be already existing
+      System.out.println("DO NOT Create table and write JSON Text " + createDefaultTable);
+      writeApiRes =
+          this.bigQueryTemplate.writeJsonStream(
+              tableName, new ByteArrayInputStream(jsonRows.getBytes()));
+    }
+    return getWriteApiResponse(writeApiRes, tableName);
+  }
+
+  private ModelAndView getWriteApiResponse(
+      ListenableFuture<WriteApiResponse> writeApiFuture, String tableName) {
+    String message = null;
+    try {
+      WriteApiResponse apiResponse = writeApiFuture.get();
+      if (apiResponse.isSuccessful()) {
+        message = "Successfully loaded data to " + tableName;
+      } else if (apiResponse.getErrors() != null && apiResponse.getErrors().size() > 0) {
```

Comment:
Following Effective Java recommendations, could the contract of `getErrors()` be modified to always return a collection -- even if that collection is empty?

Reply:
sure. it will always return a `List<StorageError>` , even if empty